File: RockitCompiler/benchmarks/bench_binarytrees.js
// bench_binarytrees.js — Binary Trees (CLBG)
// Single-threaded reference implementation

class Node {
    constructor() {
        this.left = null;
        this.right = null;
    }
}

function bottomUpTree(depth) {
    const n = new Node();
    if (depth > 0) {
        n.left = bottomUpTree(depth - 1);
        n.right = bottomUpTree(depth - 1);
    }
    return n;
}

function itemCheck(node) {
    if (node.left === null) return 1;
    return 1 + itemCheck(node.left) + itemCheck(node.right);
}

function main() {
    const n = 21;
    const minDepth = 4;
    const maxDepth = (minDepth + 2 > n) ? minDepth + 2 : n;
    const stretchDepth = maxDepth + 1;

    const stretchTree = bottomUpTree(stretchDepth);
    console.log("stretch tree of depth " + stretchDepth + "\t check: " + itemCheck(stretchTree));

    const longLivedTree = bottomUpTree(maxDepth);

    for (let depth = minDepth; depth <= maxDepth; depth += 2) {
        const iterations = 1 << (maxDepth - depth + minDepth);
        let check = 0;
        for (let i = 0; i < iterations; i++) {
            const tree = bottomUpTree(depth);
            check += itemCheck(tree);
        }
        console.log(iterations + "\t trees of depth " + depth + "\t check: " + check);
    }

    console.log("long lived tree of depth " + maxDepth + "\t check: " + itemCheck(longLivedTree));
}

main();

File: RockitCompiler/benchmarks/bench_binarytrees.rs
// bench_binarytrees.rs — Binary Trees (CLBG)
// Single-threaded reference implementation

struct Node {
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

fn bottom_up_tree(depth: i32) -> Box<Node> {
    if depth > 0 {
        Box::new(Node {
            left: Some(bottom_up_tree(depth - 1)),
            right: Some(bottom_up_tree(depth - 1)),
        })
    } else {
        Box::new(Node { left: None, right: None })
    }
}

fn item_check(node: &Node) -> i64 {
    match (&node.left, &node.right) {
        (Some(l), Some(r)) => 1 + item_check(l) + item_check(r),
        _ => 1,
    }
}

fn main() {
    let n = 21;
    let min_depth = 4;
    let max_depth = if min_depth + 2 > n { min_depth + 2 } else { n };
    let stretch_depth = max_depth + 1;

    let stretch_tree = bottom_up_tree(stretch_depth);
    println!("stretch tree of depth {}\t check: {}", stretch_depth, item_check(&stretch_tree));
    drop(stretch_tree);

    let long_lived_tree = bottom_up_tree(max_depth);

    let mut depth = min_depth;
    while depth <= max_depth {
        let iterations = 1 << (max_depth - depth + min_depth);
        let mut check: i64 = 0;
        for _ in 0..iterations {
            let tree = bottom_up_tree(depth);
            check += item_check(&tree);
        }
        println!("{}\t trees of depth {}\t check: {}", iterations, depth, check);
        depth += 2;
    }

    println!("long lived tree of depth {}\t check: {}", max_depth, item_check(&long_lived_tree));
}
